pkg/tabulator/tabstate.go

```diff
@@ -276,50 +269,70 @@ func TabStatePath(configPath gcs.Path, tabPrefix, dashboardName, tabName string)
 	return np, nil
 }
 
-func tabulate(ctx context.Context, client gcs.Client, cfg *configpb.DashboardTab, testGroupPath, tabStatePath gcs.Path, confirm bool) error {
-	r, _, err := client.Open(ctx, testGroupPath)
+func tabulate(ctx context.Context, client gcs.Client, cfg *configpb.DashboardTab, testGroupPath, tabStatePath gcs.Path, confirm, dropEmptyCols bool) error {
+	rawGrid, _, err := gcs.DownloadGrid(ctx, client, testGroupPath)
 	if err != nil {
-		return fmt.Errorf("client.Open(%s): %w", testGroupPath.String(), err)
+		return fmt.Errorf("downloadGrid: %w", err)
 	}
-	defer r.Close()
-	z, err := zlib.NewReader(r)
-	if err != nil {
-		return fmt.Errorf("zlib.NewReader: %w", err)
-	}
-	defer z.Close()
-	buf, err := ioutil.ReadAll(z)
+
+	filterRows, err := filterGrid(cfg.GetBaseOptions(), rawGrid.GetRows())
 	if err != nil {
-		return fmt.Errorf("ioutil.ReadAll: %w", err)
-	}
-	var g statepb.Grid
-	if err = proto.Unmarshal(buf, &g); err != nil {
-		return fmt.Errorf("proto.Unmarshal: %w", err)
+		return fmt.Errorf("filterGrid: %w", err)
 	}
+	rawGrid.Rows = filterRows
 
-	newRows, err := filterGrid(cfg.GetBaseOptions(), g.GetRows())
+	// TODO(chases2): Instead of inflate/drop/rewrite, move to inflate/drop/append
+	grid, _, err := updater.InflateGrid(ctx, rawGrid, time.Time{}, time.Now())
 	if err != nil {
-		return fmt.Errorf("filterGrid: %w", err)
+		return fmt.Errorf("inflateGrid: %w", err)
+	}
+	if dropEmptyCols {
+		grid = dropEmptyColumns(grid)
 	}
-	g.Rows = newRows
 
 	if confirm {
-		buf, err = proto.Marshal(&g)
-		if err != nil {
-			return fmt.Errorf("proto.Marshal: %w", err)
+		var newGrid statepb.Grid
+		rows := map[string]*statepb.Row{} // For fast target => row lookup
+
+		for _, col := range grid {
+			// TODO(chases2): refactor updater.ConstructGrid so that the tabulator also sets alerts, but not other things
+			updater.AppendColumn(&newGrid, rows, col)
 		}
 
-		var zbuf bytes.Buffer
-		zw := zlib.NewWriter(&zbuf)
-		_, err = zw.Write(buf)
+		buf, err := gcs.MarshalGrid(&newGrid)
 		if err != nil {
-			return fmt.Errorf("zlib.Write: %w", err)
+			return fmt.Errorf("marshalGrid: %w", err)
 		}
-		zw.Close()
 
-		_, err = client.Upload(ctx, tabStatePath, zbuf.Bytes(), false, "")
+		_, err = client.Upload(ctx, tabStatePath, buf, false, "")
 		if err != nil {
 			return fmt.Errorf("client.Upload(%s): %w", tabStatePath.String(), err)
 		}
 	}
 	return nil
 }
+
+// dropEmptyColumns drops every column in-place that has no results
+func dropEmptyColumns(grid []updater.InflatedColumn) []updater.InflatedColumn {
+	result := make([]updater.InflatedColumn, 0, len(grid))
+	lastIndex := 0
+	for index, col := range grid {
+		drop := true
+		for _, cell := range col.Cells {
+			if cell.Result != test_status.TestStatus_NO_RESULT {
+				drop = false
+				break
+			}
+		}
+		if drop {
+			result = append(result, grid[lastIndex:index]...)
```

Comment:
I might be missing something, but is there a reason this appends multiple things at once instead of appending a single column at a time? (I think the latter might be more readable, e.g. for each col, append to result if it's non-empty).

Reply:
Fewer append() calls, but more complexity and local variables. Not sure it's worth it. WDYT of this version?

Reply:
Oh that's even simpler than I was thinking, looks great! ^^